### PR TITLE
feat: add user-driven sidebar links via config

### PR DESCRIPTION
**Pull request by ca-scribner**
### Summary

This PR extends the dynamic sidebar from #118 to allow for user-defined links to be specified via config.  

### Usage/Testing

To set this config, use an intermediate yaml file:

sidebar.yaml
```
- icon: assessment
  link: /1
  text: '1'
  type: item
- icon: assessment
  link: /2
  text: '2'
  type: item
```

```bash
juju config kubeflow-dashboard additional-sidebar-links=@sidebar.yaml
```

Result: 
image

To edit the config, do:
```bash
juju config kubeflow-dashboard additional-sidebar-links > sidebar2.yaml
# edit sidebar2.yaml to add/remove links
juju config kubeflow-dashboard additional-sidebar-links=@sidebar2.yaml
```

Can also be used in conjunction with #118's application-driven links (see #118)

# Todo before final review

* [x] add usage instructions to readme.md --> edit: covered in #126 

Merge after #118

### CODE REVIEW

**Comment by beliaev-maksim** [comment]
just to confirm, did not we want to allow users to shuffle stuff and add new things. While core functionality should not be removed?

eg user should not be able to overwrite pipelines, notebooks, etc

**Comment by kimwnasptd** [comment]
My understanding from the PR is the following, but @ca-scribner please confirm:

The CentralDashboard now will populate links by aggregating the links defined:
1. In the CentralDashboard's config (`juju config ...`)
2. From the relations to the other charms, and this information is extracted with the common library implemented in https://github.com/canonical/charm-relation-interfaces/pull/83

By default the CentralDashboard is not configured with any links. So this information will be filled from the Charms that will be related to the CentralDashboard, using the library. 

If an admin would like some custom links, that don't respond to deployed Charms, then they should use the CentralDashboard's config.

Lastly admins can configure the ordering from yet another config in the CentralDashboard. This config will have a default value of: `'["Notebooks", "Experiments (AutoML)", "Experiments (KFP)", "Pipelines", "Runs", "Recurring Runs", "Volumes", "Tensorboards"]'`

**Comment by beliaev-maksim** [comment]
@kimwnasptd so it will be Append instead of replace ?

**Comment by kimwnasptd** [comment]
Yes, it should be append. Since we want the links from the related charms to always be visible.

But if a user would like some extra link, they could do it with config of the CentralDashboard.

**Comment by ca-scribner** [comment]
This function is described more in [this spec](https://docs.google.com/document/d/1-V34zue6qLlGr1zRfQk3R9EkeF2iGTrQr1fmsv8pEtI/edit), so whatever we decide we should make sure it is clear there.  

@kimwnasptd has it exactly right.  Stating another way just in case it isn't clear, the dashboard charm by default has no sidebar links and sidebar links can be added by:
* relating a charm to dashboard
* putting a link in the `additional-sidebar-links` config

Removal of links is done by the reverse of the above two (removing a relation, or removing a link from the config).

File: tests/integration/test_charm.py [hunk 1]
```diff
@@ -143,8 +148,8 @@ async def test_configmap_contents_with_relations(
     ops_test: OpsTest, copy_libraries_into_tester_charm, lightkube_client: Client
 ):
     """Tests the contents of the dashboard sidebar link configmap when relations are present."""
-    tester1 = "kubeflow-dashboard-requirer-tester1"
-    tester2 = "kubeflow-dashboard-requirer-tester2"
+    tester1 = f"{TESTER_CHARM_NAME}1"
+    tester2 = f"{TESTER_CHARM_NAME}2"
     charm = await ops_test.build_charm("./tests/integration/sidebar_requirer_tester_charm")
```
Comment (NohaIhab) [inline tests/integration/test_charm.py:153]: Can you make it clear in the docstring of  the test that we're using this test charm and what it does (mock charm?), I had no idea what it is and went to look at the files there

File: src/charm.py [hunk 2]
```diff
@@ -233,6 +240,51 @@ def _deploy_k8s_resources(self) -> None:
...
+            f"`{ADDITIONAL_SIDEBAR_LINKS_CONFIG}` - ignoring this input."
+        )
+
+        sidebar_config = self.model.config[ADDITIONAL_SIDEBAR_LINKS_CONFIG]
+        if not sidebar_config:
+            return []
+
+        try:
+            user_sidebar_links = yaml.safe_load(sidebar_config)
+        except Exception as err:  # TODO: Make this the right error
```
Comment (NohaIhab) [inline src/charm.py:276]: would this be yaml.YAMLError ?

File: tests/integration/test_charm.py (hunk 1 above)
Comment (ca-scribner) [reply tests/integration/test_charm.py:153]: added!

File: src/charm.py (hunk 2 above)
Comment (ca-scribner) [reply src/charm.py:276]: good call!  Didn't realise that is a parent to all the errors it could raise.  ty!

**Pull request merged**